L1Trigger/TrackFindingTMTT/src/Histos.cc

```diff
@@ -0,0 +1,3971 @@
+      }
+    }
+    for (unsigned int iNonant = 0; iNonant < numPhiNonants; iNonant++) {
+      hisStubsOnTracksPerNon_[tName]->Fill(nStubsOnTracksInNonant[iNonant]);
+      // Plot number of stubs assigned to tracks per nonant, never counting each individual stub more than once.
+      hisUniqueStubsOnTrksPerNon_[tName]->Fill(uniqueStubsOnTracksInNonant[iNonant].size());
+    }
+
+    // Plot number of tracks & number of stubs per output HT opto-link.
+
+    if (TMTT && not withRZfilter) {
+      static std::atomic<bool> firstMess = true;
+      const unsigned int numPhiSecPerNon = numPhiSectors_ / numPhiNonants;
+      // Hard-wired bodge
+      const unsigned int nLinks = houghNbinsPt_ / 2;  // Hard-wired to number of course HT bins. Check.
+
+      for (unsigned int iPhiNon = 0; iPhiNon < numPhiNonants; iPhiNon++) {
+        // Each nonant has a separate set of links.
+        vector<unsigned int> stubsToLinkCount(nLinks, 0);  // Must use vectors to count links with zero entries.
+        vector<unsigned int> trksToLinkCount(nLinks, 0);
+        for (const L1track3D& trk : tracks) {
+          unsigned int iNonantTrk = floor((trk.iPhiSec()) * numPhiNonants / (numPhiSectors_));  // phi nonant number
+          if (iPhiNon == iNonantTrk) {
+            unsigned int link = trk.optoLinkID();
+            if (link < nLinks) {
+              stubsToLinkCount[link] += trk.getNumStubs();
+              trksToLinkCount[link] += 1;
+            } else if (firstMess) {
+              firstMess = false;
+              cout << endl << "===== HISTOS MESS UP: Increase size of nLinks ===== " << link << endl << endl;
+            }
+          }
+        }
+
+        for (unsigned int link = 0; link < nLinks; link++) {
+          unsigned int nstbs = stubsToLinkCount[link];
+          hisNumStubsPerLink_[tName]->Fill(nstbs);
+          hisNumStubsVsLink_[tName]->Fill(link, nstbs);
+          profMeanStubsPerLink_[tName]->Fill(link, nstbs);
+        }
+
+        for (unsigned int link = 0; link < nLinks; link++) {
+          unsigned int ntrks = trksToLinkCount[link];
+          hisNumTrksPerLink_[tName]->Fill(ntrks);
+          hisNumTrksVsLink_[tName]->Fill(link, ntrks);
+          profMeanTrksPerLink_[tName]->Fill(link, ntrks);
+        }
+      }
+    }
+
+    // Plot q/pt spectrum of track candidates, and number of stubs/tracks
+    for (const L1track3D& trk : tracks) {
+      hisNumTracksVsQoverPt_[tName]->Fill(trk.qOverPt());  // Plot reconstructed q/Pt of track cands.
+      hisStubsPerTrack_[tName]->Fill(trk.getNumStubs());   // Stubs per track.
+      const TP* tp = trk.getMatchedTP();
+      if (TMTT) {
+        // For genuine tracks, check how often they have too many stubs to be stored in cell memory. (Perhaps worse for high Pt particles in jets?).
+        if (tp != nullptr) {
+          if (tp->useForAlgEff())
+            profExcessStubsPerTrackVsPt_[tName]->Fill(1. / tp->pt(), trk.getNumStubs() > 16);
```

Comment:
Only a dummy version of class Histos, with no implementation, will be included in the updated PR. (This will allow us to replace it will the full implementation for single-threaded L1 track development work, but will avoid the problem of histograms preventing multithreaded EDProducers for CMSSW official code)